source/tools/glkernel-cli/PngExporter.cpp

```diff
@@ -0,0 +1,210 @@
+  png_save_uint_16(outputRow + 8 * x + 2, static_cast<uint16_t>(cellValue.y));
+  png_save_uint_16(outputRow + 8 * x + 4, static_cast<uint16_t>(cellValue.z));
+  png_save_uint_16(outputRow + 8 * x + 6, static_cast<uint16_t>(cellValue.w));
+}
+
+void PngExporter::exportKernel() {
+    png_bytepp pngData;
+    int colorType;
+    png_uint_32 width;
+    png_uint_32 height;
+    if (m_kernel.hasType<glkernel::kernel4>())
+    {
+        colorType = PNG_COLOR_TYPE_RGBA;
+        const auto kernel = m_kernel.value<glkernel::kernel4>();
+        width = kernel.width();
+        height = kernel.height();
+        pngData = toPng(kernel, 4);
+    }
+    else if (m_kernel.hasType<glkernel::kernel3>())
+    {
+        colorType = PNG_COLOR_TYPE_RGB;
+        const auto kernel = m_kernel.value<glkernel::kernel3>();
+        width = kernel.width();
+        height = kernel.height();
+        pngData = toPng(kernel, 3);
+    }
+    else if (m_kernel.hasType<glkernel::kernel2>())
+    {
+        colorType = PNG_COLOR_TYPE_GA;
+        const auto kernel = m_kernel.value<glkernel::kernel2>();
+        width = kernel.width();
+        height = kernel.height();
+        pngData = toPng(kernel, 2);
+    }
+    else if (m_kernel.hasType<glkernel::kernel1>())
+    {
+        colorType = PNG_COLOR_TYPE_GRAY;
+        const auto kernel = m_kernel.value<glkernel::kernel1>();
+        width = kernel.width();
+        height = kernel.height();
+        pngData = toPng(kernel, 1);
+    }
+    else
+    {
+        cppassist::error() << "Unknown kernel type found. Aborting...";
+        return;
+    }
+
+    // TODO multiply height and width with depth for 3D kernels
+    writeToFile(pngData, colorType, height, width);
+}
+
+
+// TODO support 3D kernels with depth
+// https://github.com/p-otto/glkernel/issues/44
+template <typename T>
+png_bytepp PngExporter::toPng(const glkernel::tkernel<T> & kernel, const int channels)
+{
+    const auto minmax = findMinMaxElements(kernel);
+    const auto min = minmax.first;
+    const auto max = minmax.second;
+
+    // TODO calculate scaling factor, error rate, variance, stddeviation
+    // https://github.com/p-otto/glkernel/issues/47
+    cppassist::info() << "Scaling floating point range [" << min << ", " << max << "] to integer range [0, 65536]";
+
+
+    // memory for all rows:
+    auto rows = (png_bytepp) malloc(kernel.height() * sizeof(png_bytep));
+    // memory for one row: amount of channes * amount of pixels * png byte size
+    for(int y = 0; y < kernel.height(); ++y) {
+        // we are using 16 bit depth, so we need 2 bytes per channel
+        rows[y] = (png_bytep) malloc(channels * kernel.width() * 2 * sizeof(png_byte));
+    }
+
+    for (auto y = 0; y < kernel.height(); ++y)
+    {
+        for (auto x = 0; x < kernel.width(); ++x)
+        {
+            auto value = kernel.value(x, y, 0);
+            auto normalizedValue = (value - min) / (max - min);
```

Comment:
There is no need to recompute the difference during each iteration.
I actually hope the compiler removes the recomputations, but still...